src/sample.rs

```diff
@@ -0,0 +1,330 @@
+    fn sample<R: Rng>(&self, rng: &mut R, amount: usize) -> Self::Sampled;
+}
+
+/// The `SampleRef` trait provides the `sample_ref` method.
+///
+/// This is intended to be implemented for containers that which can be sampled in `O(amount)` time
+/// and want a fast way to give references to a sample of their items.
+pub trait SampleRef {
+    /// The returned sampled data. Typically the either a `Vec<&T>` or a new instance of the
+    /// container's own type containing references to the underlying data.
+    type SampledRef;
+
+    /// Return exactly `amount` references to randomly sampled values.
+    ///
+    /// Any type which implements `sample_ref` should guarantee that:
+    /// - Both the order and values of `SampledRef` is random.
+    /// - The implementation uses `O(amount)` speed and memory.
+    /// - The returned values are not copies/clones (if so, implement `Sample` instead).
+    ///
+    /// Panics if `amount > self.len()`
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use rand::{thread_rng, SampleRef};
+    ///
+    /// let mut rng = thread_rng();
+    /// let values = vec![5, 6, 1, 3, 4, 6, 7];
+    /// println!("{:?}", values.as_slice().sample_ref(&mut rng, 3))
+    /// ```
+    fn sample_ref<R: Rng>(&self, rng: &mut R, amount: usize) -> Self::SampledRef;
+}
+
+/// Randomly sample *up to* `amount` elements from a finite iterator using a reservoir.
+///
+/// The order of elements in the sample is not random. In fact, if `len(iterable) <= amount` then
+/// the output will be in the exact order they were collected.
+///
+/// The reservoir method used allocates only an `Vec` of size `amount`. The size of the iterable
+/// does not affect the amount of memory used.
+///
+/// # Example
+///
+/// ```rust
+/// use rand::{thread_rng, sample_reservoir};
+///
+/// let mut rng = thread_rng();
+/// let sample = sample_reservoir(&mut rng, 1..100, 5);
+/// println!("{:?}", sample);
+/// ```
+pub fn sample_reservoir<T, I, R>(rng: &mut R, iterable: I, amount: usize) -> Vec<T>
+    where I: IntoIterator<Item=T>,
+          R: Rng,
+{
+    let mut iter = iterable.into_iter();
+    let mut reservoir = Vec::with_capacity(amount);
+    reservoir.extend(iter.by_ref().take(amount));
+
+    // continue unless the iterator was exhausted
+    if reservoir.len() == amount {
+        for (i, elem) in iter.enumerate() {
+            let k = rng.gen_range(0, i + 1 + amount);
+            if let Some(spot) = reservoir.get_mut(k) {
+                *spot = elem;
+            }
+        }
+    }
+    reservoir
+}
+
+/// Sample (non-repeating) exactly `amount` of indices from a sequence of the given `length`.
+///
+/// The returned elements and their order are random.
+///
+/// Panics if `amount > length`
+///
+/// TODO: IMO this should be made public since it can be generally useful, although
+/// there might be a way to make the output type more generic/compact.
+fn sample_indices<R>(rng: &mut R, length: usize, amount: usize) -> Vec<usize>
```

Comment:
So make it public? Prototype seems okay. Maybe it should say: samples *amount* numbers from the range `0..length`.

Reply:
I think I can make it more general, `sample_range` to sample over *any* numeric range, and even make the input types generic. I'd like to do that before exporting it.

I'll open a ticket once this gets merged. It's out of scope and just a "nice to have".